crates/crypto/addresses/src/address/decl_address.rs

```diff
@@ -61,25 +54,70 @@ macro_rules! decl_ret_wrapped_address {
-                derive_more::Display,
                 derive_more::Debug,
                 SerializeDisplay,
                 DeserializeFromStr,
             )]
-            #[display("{}", self.0)]
-            #[debug("{}", self.0)]
-            pub struct [< $address_type:camel Address >](pub(crate) [< Ret $address_type:camel Address >]);
+            #[debug("{}", self)]
+            pub struct [< $address_type:camel Address >] {
+                /// The NodeId of the address.
+                node_id: ScryptoNodeId,
+                /// The network that the address is to be used for. This is
+                /// used in the Bech32m encoding and decoding of the address
+                /// essentially providing us with the network context.
+                network_id: NetworkID,
+                /// The entity type of the address. This is checked in the
+                /// constructor that it is one of the allowed entity types
+                /// for this particular address type and then cached here
+                /// to avoid any additional unwraps when being retrieved.
+                entity_type: ScryptoEntityType
+            }
 
-            impl From<[< Ret $address_type:camel Address >]> for [< $address_type:camel Address >] {
-                fn from(value: [< Ret $address_type:camel Address >]) -> Self {
-                    Self(value)
+            impl AddressFromNodeId for [< $address_type:camel Address >] {
+                fn new_from_node_id(
+                    node_id: impl Into<ScryptoNodeId>,
+                    network_id: NetworkID
+                ) -> Result<Self> {
+                    let node_id = node_id.into();
+                    if let Some(entity_type) = node_id.entity_type() {
+                        if matches!(
+                            entity_type,
+                            $($entity_type)|*
+                        ) {
+                            Ok(Self {
+                                node_id,
+                                network_id,
+                                entity_type
+                            })
+                        } else {
+                            Err(CommonError::AddressInvalidEntityType { address_kind: stringify!([< $address_type:camel >]).to_string(), entity_type: entity_type as u8, node_id_as_hex: node_id.to_hex() })
+                        }
+                    } else {
+                        Err(CommonError::AddressNodeIdNoEntityType { node_id_as_hex: node_id.to_hex() })
+                    }
                 }
             }
 
-            impl FromStr for [< $address_type:camel Address >] {
-                type Err = CommonError;
-                fn from_str(s: &str) -> Result<Self> {
-                    Self::try_from_bech32(s)
+            impl Display for [< $address_type:camel Address >] {
+                fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                    let encoder = AddressBech32Encoder::new(
```

Comment:
possibly we will wanna do this **once** in ctor, and save it as a field? Memory cost (and UniFFI roundtrip performance cost) of sliiiighly larger types < performance cost of calling this many times? Not sure, but maybe!